lib/system/iterators.nim

```diff
@@ -129,14 +129,141 @@ iterator items*[T: Ordinal](s: Slice[T]): T =
   for x in s.a .. s.b:
     yield x
 
+iterator ritems*[T: not char](a: openArray[T]): lent2 T {.inline.} =
+  ## Iterates over each item of `a` in reverse.
+  var i = high(a)
+  while i >= low(a):
+    yield a[i]
+    dec(i)
+
+iterator ritems*[T: char](a: openArray[T]): T {.inline.} =
+  ## Iterates over each item of `a` in reverse.
+  # a VM bug currently prevents taking address of openArray[char]
+  # elements converted from a string (would fail in `tests/misc/thallo.nim`)
+  # in any case there's no performance advantage of returning char by address.
+  var i = high(a)
+  while i >= low(a):
+    yield a[i]
+    dec(i)
+
+iterator mritems*[T](a: var openArray[T]): var T {.inline.} =
+  ## Iterates over each item of `a` in reverse so that the yielded value can be modified.
+  var i = high(a)
+  while i >= low(a):
+    yield a[i]
+    dec(i)
+
+iterator ritems*[IX, T](a: array[IX, T]): T {.inline.} =
+  ## Iterates over each item of `a` in reverse.
+  when a.len > 0:
+    var i = high(IX)
+    while true:
+      yield a[i]
+      if i < low(IX): break
+      dec(i)
+
+iterator mritems*[IX, T](a: var array[IX, T]): var T {.inline.} =
+  ## Iterates over each item of `a`in reverse so that the yielded value can be modified.
+  when a.len > 0:
+    var i = high(IX)
+    while true:
+      yield a[i]
+      if i < low(IX): break
+      dec(i)
+
+iterator ritems*[T](a: set[T]): T {.inline.} =
+  ## Iterates over each element of `a` in reverse. `items` iterates only over the
+  ## elements that are really in the set (and not over the ones the set is
+  ## able to hold).
+  var i = high(T).int
+  while i > low(T).int:
+    if T(i) in a: yield T(i)
+    dec(i)
```

Comment:
I'm not sure this necessarily makes sense. The elements are ordered by their value, but that's more of an implementation detail (a `set[T]` is a bitset), since a set usually is an unordered collection of elements.

Reply:
But if you iterate over a set it's sorted from the lowest bit to the highest so if you want to do it in reverse you must use this, anyways I need to test if it works (and add examples to all the other templates).
```nim
for i in {1, 2}:
  echo i # 1 2

for i in {1, 2}.ritems: # or {2, 1}.ritems
  echo i # 2 1
```